stale_read_user_guide.md

```diff
@@ -0,0 +1,186 @@
+
+### Resolved-ts: What Is It?
+
+Like safe-ts, resolved-ts is a timestamp that guarantees all transactions with a timestamp less than this value have been applied by the leader. It is different from safe-ts in that safe-ts is a per peer concept, while resolved-ts is only maintained by the leader. Followers can have a smaller apply index than the leader, so resolved-ts cannot be directly treated as safe-ts in followers.
+
+### The Maintenance of Resolved-ts
+
+Region leaders manage resolved-ts using a resolver. This resolver tracks locks in the LOCK CF(Column Family) by receiving change logs when Raft applies. When initialized, the resolver scans the entire region to track the locks.
+
+## Diagnosing Issues with Stale Read
+
+### Identifying Issues
+
+In Grafana - TiDB dashboard - KV request row, there are panels showing the hit rate, OPS and traffic of stale read.
+
+![img](/media/stale-read/metrics-hit-miss.png)
+
+![img](/media/stale-read/metrics-ops.png)
+
+![img](/media/stale-read/traffic.png)
+
+The detailed explanation of these metrics can be found in [TiDB Monitoring Metrics](/grafana-tidb-dashboard.md#kv-request)
+
+When stale read encounters problems, changes in these p may be noticeable. The most direct evidence is the WARN log from TiDB, which reports 'DataIsNotReady' along with the region id and `safe-ts` it encounters.
+
+### Commonly Seen Causes
+
+The most typical issues that can hinder the utilization of stale read include:
+
+- Slow-committing ongoing transactions.
+- Delays in the CheckLeader's information push from the leader to the follower.
+
+### Using Grafana for Diagnosis
+
+In the tikv-details dashboard, within the resolved-ts row, you can identify the region with the smallest resolved-ts and safe-ts in each TiKV. If these values lag significantly behind real-time, these regions' details should be checked via `tikv-ctl`.
+
+### Using tikv-ctl for Diagnosis
+
+`tikv-ctl` can provide up-to-date details of the resolver and `RegionReadProgress`. For details, check [tikv-control](/tikv-control.md#get-the-state-of-a-regions-regionreadprogress).
+
+An example usage
+
+```
+./tikv-ctl --host 127.0.0.1:20160 get-region-read-progress -r 14 --log --min-start-ts 0
+```
+
+The output of this command might look like:
+
+```
+Region read progress:
+    exist: true,
+    safe_ts: 0,
+    applied_index: 92,
+    pending front item (oldest) ts: 0,
+    pending front item (oldest) applied index: 0,
+    pending back item (latest) ts: 0,
+    pending back item (latest) applied index: 0,
+    paused: false,
+Resolver:
+    exist: true,
+    resolved_ts: 0,
+    tracked index: 92,
+    number of locks: 0,
+    number of transactions: 0,
+    stopped: false,
+```
+
+This information can help determine:
+
+- Whether locks are blocking resolved-ts.
+- Whether the apply index is too small, preventing safe-ts from being updated.
+- If this is a follower peer, whether the leader is sending a sufficiently new resolved-ts.
```

Comment:
````suggestion
### Use `tikv-ctl` to diagnose

`tikv-ctl` provides up-to-date details of the resolver and `RegionReadProgress`. For more details, see [Get the state of a Region's `RegionReadProgress`](/tikv-control.md#get-the-state-of-a-regions-regionreadprogress).

The following is an example:

```bash
./tikv-ctl --host 127.0.0.1:20160 get-region-read-progress -r 14 --log --min-start-ts 0
```

The output is as follows:

```log
Region read progress:
    exist: true,
    safe_ts: 0,
    applied_index: 92,
    pending front item (oldest) ts: 0,
    pending front item (oldest) applied index: 0,
    pending back item (latest) ts: 0,
    pending back item (latest) applied index: 0,
    paused: false,
Resolver:
    exist: true,
    resolved_ts: 0,
    tracked index: 92,
    number of locks: 0,
    number of transactions: 0,
    stopped: false,
```

The preceding output helps you determine:

- Whether locks are blocking resolved-ts.
- Whether the apply index is too small to update safe-ts.
- Whether the leader is sending a sufficiently updated resolved-ts when a follower peer exists.
````